Add helper method to quickly create run template from pipeline

schustmi opened the pull request:
## Describe changes

This PR adds a small helper method to create a run template for a pipeline without running it first:
```python

@pipeline
def my_pipeline():
  ...

template = my_pipeline.create_run_template(name="template_name")
```

This can be useful in CI scenarios, where only a template should be created for later runs instead of running it immediately.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

bcdurak reviewed (approved):
When I tried to test this, I ran into some stuff that I was not expecting: 

1. I forgot to switch to a remote stack and I got this error:

  `ValueError: Unable to create run template as there is no associated build.`
  
  While I understand where this is coming from, the new users might be a bit confused due to the message here. Maybe we can add something like: `Make sure you have an active remote stack.`. 

2. When I tried to use the same name for the template, I got the error:

`EntityExistsError: Unable to create run template in workspace 'default': A run template with this name already exists.`
  
I think this is more of a general issue, but I would remove the mention of workspaces here and add the name of the entity to the message.

3. I also realized one more thing about the run templates. I had some `DockerSettings` defined for my pipeline and before we could create the template, I had to create and push the build. When I look at the editor on the dashboard, I see the `DockerSettings` are not there, as the template is already tied to a build. I understand this is what's needed atm, but it might be nice to adjust things like builds or stacks of the templates in the future.

The issues that I mentioned are minor issues, feel free to merge if you want.

schustmi commented on src/zenml/steps/utils.py:
```diff
@@ -538,6 +538,7 @@ def single_step_pipeline() -> None:
         __step(**inputs)
 
     run = single_step_pipeline.with_options(unlisted=True)()
+    assert run
```
I updated the return type of `Pipeline.__call__` to be an `Optional[PipelineRunResponse]` instead of `Any`. This enables autocomplete on the run object. However, as this is optional (`None` in case of scheduled runs), I need this assert for mypy.

schustmi commented:
> When I tried to test this, I ran into some stuff that I was not expecting:
> 
> 1. I forgot to switch to a remote stack and I got this error:
> 
> `ValueError: Unable to create run template as there is no associated build.`
> 
> While I understand where this is coming from, the new users might be a bit confused due to the message here. Maybe we can add something like: `Make sure you have an active remote stack.`.
> 
> 2. When I tried to use the same name for the template, I got the error:
> 
> `EntityExistsError: Unable to create run template in workspace 'default': A run template with this name already exists.`
> 
> I think this is more of a general issue, but I would remove the mention of workspaces here and add the name of the entity to the message.
> 
> 3. I also realized one more thing about the run templates. I had some `DockerSettings` defined for my pipeline and before we could create the template, I had to create and push the build. When I look at the editor on the dashboard, I see the `DockerSettings` are not there, as the template is already tied to a build. I understand this is what's needed atm, but it might be nice to adjust things like builds or stacks of the templates in the future.
> 
> The issues that I mentioned are minor issues, feel free to merge if you want.

1. Some remote orchestrators are not supported, only containerized ones are. Which is why we need to check for a build here. I've updated the error message to try to make it a little clearer though
2. I added the template name to the error message. I think the workspace stuff I'd leave in there for now, until we either remove or rework them, as it's in 100 other places as well.
3. 100% agree, but this would be months of effort to get this working I think :P
